plugins/multiconnect.c

```diff
@@ -0,0 +1,218 @@
+	 * for this connection.  */
+	idtok = json_get_member(buf, result, "id");
+	if (!idtok)
+		plugin_err(cmd->plugin, "'connect' missing 'id' field");
+
+	/* Replace whatever string was originally in here.  */
+	tal_free(css->id);
+	css->id = json_strdup(cmd, buf, idtok);
+
+	/* Retrieve the `features` from the base command.  */
+	featurestok = json_get_member(buf, result, "features");
+	if (!featurestok)
+		plugin_err(cmd->plugin, "'connect' missing 'features' field");
+	css->features = json_strdup(cmd, buf, featurestok);
+
+	return plugin_spark_complete(cmd, css->completion);
+}
+static struct command_result *
+connect_single_spark_start(struct command *cmd,
+			   struct plugin_spark_completion *completion,
+			   struct connect_single_spark *css)
+{
+	struct out_req *req;
+
+	css->completion = completion;
+
+	req = jsonrpc_request_start(cmd->plugin, cmd, "connect",
+				    &connect_single_spark_done,
+				    &forward_error,
+				    css);
+	json_add_string(req->js, "id", css->id);
+	return send_outreq(cmd->plugin, req);
+}
+
+static struct command_result *multiconnect(struct command *cmd,
+					   const char *buf,
+					   const jsmntok_t *ids)
+{
+	size_t i;
+	const jsmntok_t *t;
+	struct connect_multi *cm;
+	struct plugin_spark_group *sparks;
+
+	cm = tal(cmd, struct connect_multi);
+	cm->subcommands = tal_arr(cm, struct connect_single_spark, ids->size);
+
+	sparks = plugin_make_spark_group(cm, cmd);
+
+	/* We know at this point that ids is an array of strings.  */
+	json_for_each_arr(i, t, ids) {
+		struct plugin_spark *spark;
+		cm->subcommands[i].id = json_strdup(cm, buf, t);
+		spark = plugin_start_spark(cmd,
+					   &connect_single_spark_start,
+					   &cm->subcommands[i]);
+		plugin_spark_group_add(sparks, spark);
+	}
+
+	return plugin_wait_spark_group(cmd, sparks,
+				       &multiconnect_done, cm);
+}
+
+static struct command_result *json_connect(struct command *cmd,
+					   const char *buf,
+					   const jsmntok_t *params)
+{
+	const jsmntok_t *idtok;
+	const char *host;
+	u32 *port;
+	struct out_req *req;
+
+	if (!param(cmd, buf, params,
+		   p_req("id", param_tok, &idtok),
+		   p_opt("host", param_string, &host),
+		   p_opt("port", param_number, &port),
+		   NULL))
+		return command_param_failed();
+
+	/* Is id an array?  If so use multi-connect.  */
+	if (idtok->type == JSMN_ARRAY) {
+		size_t i;
+		const jsmntok_t *t;
+
+		if (host)
+			return command_done_err(cmd, JSONRPC2_INVALID_PARAMS,
+						"Cannot specify parameter "
+						"'host' when 'id' parameter "
+						"is an array.",
+						NULL);
+		if (port)
+			return command_done_err(cmd, JSONRPC2_INVALID_PARAMS,
+						"Cannot specify parameter "
+						"'port' when 'id' parameter "
+						"is an array.",
+						NULL);
+
+		if (idtok->size == 0)
+			return command_done_err(cmd, JSONRPC2_INVALID_PARAMS,
+						"Empty 'id' array: "
+						"nothing to connect.",
+						NULL);
+
+		json_for_each_arr(i, t, idtok) {
+			if (t->type != JSMN_STRING)
```

Comment:
Maybe also check if the string length is `>= 32` so that we exit early in case of a mis-copy/paste

Reply:
The string can have an `@host:port` appended to the end.  We delegate parsing that to `connect`.  We *could* add the parsing for `@host:port` here but that would violate DRY unless we refactor the parsing code over in `lightningd/connect_control.c` and put it in a new common module.